### DISCUSSION TITLE
distanza tra due vertici

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#23] [Esercizi A-L]
Sia $G = (V, E)$ un grafo, e siano $x, y \in V$, la distanza da $x$ a $y$ è il **minimo** numero di archi in un **cammino** da $x$ a $y$ (se tale cammino non esiste, $dist(x, y) = \infty$)

Dato un grafo $G = (V, E)$ non diretto e $x, y \in V$, scrivere lo pseudocodice di un algoritmo che trova la distanza $x$ e $y$ in $O(|V|+|E|)$

**Comment by LeoCosta02** [comment]
```pseudo

BFS(G: graph , x node , y node){
    P:vettore di padri inizializzato a -1 
    int dist[n] array lungo n= |V(G)| inizializzato a 0 
    if(x==y) return 0
    P[x]=x
    Q: coda vuota
    Q.enqueue(x)
    while(Q != 0 AND P[y]==-1){
        v=Q.dequeue()
        for each(w adiacente di v ){
            if(P[w]==-1){
                Q.enque(w)
                Dist[w]=Dist[v]+1
                P[w]=v
            }
        }
    }
    if (Dist[y] ==0)
     printf("la distanza tra x e y è infinito") // x != y e se y non viene visitato 
     else
      return Dist[y]
}
```

**Comment by Frisbon** [comment]
Usando il **DFS**! (sperando funzioni lol)
Il mio ragionamento era, se vado avanti - metto sullo stack e quindi la distanza aumenta.
Se tolgo dallo stack, la distanza diminuisce. 
Esplorando tutto l'albero e lasciando $y \ (\alpha)$ non visitato, posso vedere tutte le distanze e salvarmi solo la minima

![image_2024-06-15_07-55-22](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/17a66bb9-786f-429c-af48-e1b8cef1fadd)

**Comment by daromacarlo** [comment]
```c
def BFS(G, x, y)
	dist = [infinito, ..., infinito]		//O(n) per inizializzazione
	queue = []								//coda
	dist[x] = 0
	queue.enqueue(x)
	while queue is not empty{
		v = queue.dequeue()
		for w in v.adiacenti{
			queue.enqueue(w)
			if dist[w] == infinito{
				dist[w] = (dist[v] + 1)	
			}
		}
	}
	return dist[y]


// O(N+M)
```
